src/v/kafka/security/scram_authenticator.h

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright 2020 Vectorized, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+#pragma once
+#include "kafka/security/credential_store.h"
+#include "kafka/security/sasl_authentication.h"
+#include "kafka/security/scram_algorithm.h"
+
+namespace kafka {
+
+template<typename ScramMechanism>
+class scram_authenticator final : public sasl_mechanism {
+    static constexpr int nonce_size = 130;
+
+public:
+    explicit scram_authenticator(credential_store& credentials)
+      : _state{state::client_first_message}
+      , _credentials(credentials) {}
+
+    bytes authenticate(bytes_view auth_bytes) override;
+
+    bool complete() const override { return _state == state::complete; }
+    bool failed() const override { return _state == state::failed; }
+
+    const ss::sstring& authid() const {
+        if (likely(_state == state::complete)) {
+            return _authid;
+        }
+        throw sasl_error(
```

Comment:
This looks like `std::log_error` or assert, not `sasl_error`. If I understand the patch correctly, `sasl_error` is used when the client has bad credentials or fails to comply with the auth protocol. But this error will occur when our code will invoke `authid` in the wrong way.

Reply:
sasl_error is now gone